conwayste/src/ui/button.rs

```diff
@@ -0,0 +1,187 @@
+
+use std::rc::Rc;
+
+use chromatica::css;
+
+use ggez::graphics::{self, Rect, Font, Color, DrawMode, DrawParam};
+use ggez::nalgebra::{Point2, Vector2};
+use ggez::{Context, GameResult};
+
+use super::{
+    label::Label,
+    widget::Widget,
+    helpe::{within_widget, color_with_alpha, center},
+    UIAction, WidgetID
+};
+
+pub struct Button {
+    pub id:     WidgetID,
+    pub label: Label,
+    pub button_color: Color,
+    pub draw_mode: DrawMode,
+    pub dimensions: Rect,
+    pub hover: bool,
+    pub borderless: bool,
+    pub action: UIAction
+}
+
+const BUTTON_LABEL_PADDING_W: f32 = 16.0;   // in pixels
+const BUTTON_LABEL_PADDING_H: f32 = 16.0;   // in pixels
+
+/// A named widget that can be clicked to result in an occuring action.
+impl Button {
+
+    /// Creates a Button widget.
+    ///
+    /// # Arguments
+    /// * `ctx` - GGEZ context
+    /// * `font` - font to be used when drawing the text
+    /// * `button_text` - Text to be displayed
+    /// * `widget_id` - Unique widget identifier
+    /// * `action` - Unique action identifer
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use ggez::graphics::Font;
+    /// use ui::Button;
+    ///
+    /// fn new(ctx: &mut Context) -> GameResult<MainState> {
+    ///     let font = Rc::new(Font::default());
+    ///     let b = Button::new(ctx, WidgetID::TestButton1,
+    ///             UIAction::PrintHelloWorld,
+    ///             WidgetID::TestButton1Label,
+    ///             font,
+    ///             "TestButton")
+    ///         .label_color(Color::from(css::DARKCYAN))
+    ///         .button_color(Color::from(css::WHITE));
+    ///
+    ///     b.draw(ctx)?;
+    /// }
+    /// ```
+    ///
+    pub fn new(ctx: &mut Context, widget_id: WidgetID, action: UIAction, label_id: WidgetID,
+        font: Rc<Font>, button_text: String) -> Self
+    {
+        let label_position = Point2::new(0.0, 0.0); // label positioning defined an offset to button origin after centering
+        let label = Label::new(ctx, label_id, font, button_text, color_with_alpha(css::WHITE, 0.1), label_position);
+        let label_dims = label.size();
+
+        let dimensions = Rect::new(30.0, 20.0, label_dims.w + BUTTON_LABEL_PADDING_W, label_dims.h + BUTTON_LABEL_PADDING_H);
+
+        let mut b = Button {
+            id: widget_id,
+            label: label,
+            button_color: color_with_alpha(css::DARKCYAN, 0.8),
+            draw_mode: DrawMode::fill(),
+            dimensions: dimensions,
+            hover: false,
+            borderless: false,
+            action: action,
+        };
+        b.center_label_text();
+        b
+    }
+
+    /// Sets the color of the Button's text to the specified ggez `Color`
+    pub fn label_color(mut self, color: Color) -> Self {
+        self.label.set_color(color);
+        self
+    }
+
+    /// Sets the color of the button to the specified ggez `Color`
+    pub fn button_color(mut self, color: Color) -> Self {
```

Comment:
Why have a setter? The field is public.

Reply:
Was intended to be used with the builder pattern. No use case as of yet... I know what you're going to say so I'll just go ahead and remove it.